Move thindev and thinpool from Stratis to devicemapper-rs

trgill opened the pull request:
Move thindev and thinpool from Stratis to devicemapper-rs

Minor adjustments to the use statements to match style guide
Added comments where required due to compiler warnings if not supplied
Changed fn path() to devnode() in thindev and thinpool

trgill commented:
@agrover @mulkieran I think this is a trivial move of the two files/minor adjustments

agrover commented on src/lineardev.rs:
```diff
@@ -2,12 +2,12 @@
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
-use {DM, DevId, DeviceInfo, DmFlags};
 use segment::Segment;
```
Segment should likely move down too.

agrover replied:
otherwise LGTM and 👍

trgill replied:
thanks.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    /// create an XFS file system on the block device
+    pub fn create_fs(&mut self) -> DmResult<()> {
```
Remove mut modifier for self on ```_fs``` methods.

mulkieran commented on src/thinpooldev.rs:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    /// path of the device node
+    pub fn path(&self) -> DmResult<PathBuf> {
```
should be ```devnode``` here.

mulkieran commented on src/thinpooldev.rs:
```diff
@@ -0,0 +1,118 @@
+    /// send a message to DM thin pool
+    pub fn message(&self, dm: &DM, message: &str) -> DmResult<()> {
+        try!(dm.target_msg(&DevId::Name(&self.name), 0, message));
+
```
Extra line.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,150 @@
+
+        let params = format!("{} {}", pool_dev_str, thin_id);
+
+        table.push((0u64, length.0, "thin".to_owned(), params));
```
```vec!``` would work here just as well, as there is no loop.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,150 @@
+    /// thin provisioned ThinDev returned by new().
+    pub fn new(name: &str,
+               dm: &DM,
+               thin_pool: &mut ThinPoolDev,
```
This ```mut``` can also go away.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,144 @@
+    /// "<start> <length> thin </dev/mapper/poolname> <thin_id>"
+    fn dm_table(pool_dev_str: &String,
+                thin_id: u32,
+                length: &Sectors)
```
length type should be ```Sectors```.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,144 @@
+                length: &Sectors)
+                -> Vec<(u64, u64, String, String)> {
+        let params = format!("{} {}", pool_dev_str, thin_id);
+        vec![(0u64, length.0, "thin".to_owned(), params)]
```
Use ```*length``` here.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,144 @@
+
+    /// mount the filesystem on the mount point
+    pub fn mount_fs(self, mount_point: &Path) -> DmResult<()> {
+        let path_result = try!(self.devnode());
```
These two lines can just be ```let dev_path = try!(self.devnode);```, i.e., there's no need to build a Path object. See commit: eeb0f51bd7a2ea2de78aff.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,144 @@
+
+    /// create an XFS file system on the block device
+    pub fn create_fs(self) -> DmResult<()> {
+        let path_result = try!(self.devnode());
```
These two lines can also be reduced, same as below.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    /// Remove the device from DM
+    pub fn teardown(&self, dm: &DM) -> DmResult<()> {
```
All these teardown() methods should consume ```self```, not a reference to ```self```. You could add an additional patch to this PR that makes this change for all ```teardown()``` methods.

mulkieran commented on src/thinpooldev.rs:
```diff
@@ -0,0 +1,115 @@
+
+/// DM construct to contain thin provisioned devices
+pub struct ThinPoolDev {
+    /// name
```
Either more or less on this comment would be better. Less is probably doable, since these attributes are all private. More would be nice, to help others.

trgill replied:
I replaced name with calls to name()

mulkieran commented on src/thinpooldev.rs:
```diff
@@ -0,0 +1,115 @@
+                             data_dev.to_string_lossy(),
+                             *data_block_size,
+                             *low_water_mark);
+        table.push((0u64, length.0, "thin-pool".to_owned(), params));
```
This could use a ```vec!``` construction, since their is no loop.

mulkieran replied:
```length.0``` -> ```*length```.

mulkieran commented on src/thinpooldev.rs:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    /// Remove the device from DM
+    pub fn teardown(&self, dm: &DM) -> DmResult<()> {
```
As with ```teardown()``` above and ```teardown()``` in LinearDev, these need a patch to consume self.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    /// create an XFS file system on the block device
+    pub fn create_fs(self) -> DmResult<()> {
```
Should take reference to self, ```&self```.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    /// mount the filesystem on the mount point
+    pub fn mount_fs(self, mount_point: &Path) -> DmResult<()> {
```
Should take reference to self.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    /// unmount the filesystem
+    pub fn unmount_fs(self, mount_point: &Path) -> DmResult<()> {
```
This doesn't need a ```self``` parameter at all, as far as I can tell. But, it should be left in the impl for ThinDev, I think, at least for now.

mulkieran commented on src/lineardev.rs:
```diff
@@ -88,8 +83,8 @@ impl LinearDev {
-    pub fn teardown(&self, dm: &DM) -> DmResult<()> {
-        try!(dm.device_remove(&DevId::Name(&self.name), DmFlags::empty()));
+    pub fn teardown(self, dm: &DM) -> DmResult<()> {
+        try!(dm.device_remove(&DevId::Name(&self.name()), DmFlags::empty()));
```
I think "&self.name()``` could lose the '&'.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,140 @@
+        } else {
+            let error_message = format!("{:?}", &output.stderr);
+            debug!("stderr: {}", error_message);
+            return Err(DmError::Dm(InternalError(error_message.into())));
```
These failures on ```*_fs``` methods are not internal errors. However, it's not clear to me what better options current error hierarchy has to offer. Please file an issue about this problem. The basic problem is that the error string might be a helpful thing to include in the error, but nix, which would be the natural choice, prefers just to return the error code.

trgill replied:
https://github.com/stratis-storage/devicemapper-rs/issues/15

mulkieran commented on src/thinpooldev.rs:
```diff
@@ -0,0 +1,109 @@
+
+    /// Remove the device from DM
+    pub fn teardown(self, dm: &DM) -> DmResult<()> {
+        try!(dm.device_remove(&DevId::Name(&self.name()), DmFlags::empty()));
```
I think ```&self.name()``` can lose the '&'

mulkieran commented on src/thinpooldev.rs:
```diff
@@ -0,0 +1,109 @@
+
+    /// send a message to DM thin pool
+    pub fn message(&self, dm: &DM, message: &str) -> DmResult<()> {
+        try!(dm.target_msg(&DevId::Name(&self.name()), 0, message));
```
```&self.name()``` can probably lose the '&'.

mulkieran replied:
```&self.name()``` can probably lose the '&'.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,140 @@
+
+    /// Remove the device from DM
+    pub fn teardown(self, dm: &DM) -> DmResult<()> {
+        try!(dm.device_remove(&DevId::Name(&self.name()), DmFlags::empty()));
```
```&self.name()``` probably can lose '&' here.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,140 @@
+        debug!("Create filesystem for : {:?}", dev_path);
+        let output = try!(Command::new("mkfs.xfs")
+                              .arg("-f")
+                              .arg(&dev_path)
```
This compiles w/out an '&' in next method, so I think the '&' can be removed here.

trgill replied:
you are saying &dev_path doesn't need the "&"?  If I remove it, I get a compiler error on the debug message.

mulkieran replied:
What is the compiler error?

trgill replied:
use after move

mulkieran commented on src/lineardev.rs:
```diff
@@ -41,10 +39,7 @@ impl LinearDev {
         try!(dm.device_suspend(id, DmFlags::empty()));
```
On line 35, probably "&name" can lose the '&'.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,140 @@
+
+    /// unmount the filesystem
+    pub fn unmount_fs(mount_point: &Path) -> DmResult<()> {
+        debug!("Unount filesystem {:?}", mount_point);
```
Typo.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,140 @@
+                              .output());
+
+        if output.status.success() {
+            debug!("Mounted xfs filesystem on {:?}", mount_point)
```
There is nothing about this method that makes it certain that the filesystem will be "xfs". So, that part should be left out of the debug.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,140 @@
+                              .output());
+
+        if output.status.success() {
+            debug!("Created xfs filesystem on {:?}", dev_path)
```
Leave "xfs" out.

mulkieran replied:
This particular circumstance isn't important, but I think it is wrong for logging code to affect our non-logging code. I will file an issue.

mulkieran commented on src/thindev.rs:
```diff
@@ -0,0 +1,140 @@
+use types::Sectors;
+
+/// DM construct for a thin block device
+#[derive(Clone)]
```
I don't think this needs to derive Clone.

trgill replied:
https://github.com/stratis-storage/devicemapper-rs/issues/17

agrover commented:
I think all the filesystem-related stuff should not go in thindev, it all seems pretty Stratis-specific.

trgill commented:
@agrover yeah, I should remove the filesystem related stuff from ThinDev.

I updated the first commit to do so and added the following to the commit message:
"Remove filesystem related functions since they are Stratis specific".

agrover commented:
@mulkieran Oops, I missed your comments and merged! ☹️ I've opened #19 to fix up.

mulkieran commented:
ok, thanks.
